pkg/frontend/querymiddleware/astmapper/instant_splitting.go

```diff
@@ -0,0 +1,542 @@
+		Without:  groupingWithout,
+		PosRange: parser.PositionRange{},
+	}, true, nil
+}
+
+// expr is the range vector aggregator expression
+// If the outer expression is a vector aggregator, r.embeddedAggregatorExpr will contain the expression
+// In this case, the vector aggregator should be downstream to the embedded queries in order to limit
+// the label cardinality of the parallel queries
+func (i *instantSplitter) splitAndSquashCall(expr *parser.Call, rangeInterval time.Duration) (mapped parser.Expr, finished bool, err error) {
+	splitCount := int(math.Ceil(float64(rangeInterval) / float64(i.splitByInterval)))
+	if splitCount <= 0 {
+		return expr, false, nil
+	}
+
+	var embeddedQuery parser.Expr = expr
+
+	if i.embeddedAggregatorExpr != nil {
+		embeddedQuery = i.embeddedAggregatorExpr
+	}
+
+	originalOffset := getOffset(expr)
+
+	// Create a partial query for each split
+	embeddedQueries := make([]parser.Node, 0, splitCount)
+	for split := 0; split < splitCount; split++ {
+		splitOffset := time.Duration(split) * i.splitByInterval
+		// The range interval of the last embedded query can be smaller than i.splitByInterval
+		splitRangeInterval := i.splitByInterval
+		if splitOffset+splitRangeInterval > rangeInterval {
+			splitRangeInterval = rangeInterval - splitOffset
+		}
+		// The offset of the embedded queries is always the original offset + a multiple of i.splitByInterval
+		splitOffset += originalOffset
+		splitNode, err := createSplitNode(embeddedQuery, splitRangeInterval, splitOffset)
+		if err != nil {
+			return nil, false, err
+		}
+
+		// Prepend to embedded queries
+		embeddedQueries = append([]parser.Node{splitNode}, embeddedQueries...)
+	}
+
+	squashExpr, err := vectorSquasher(embeddedQueries...)
+	if err != nil {
+		return nil, false, err
+	}
+
+	return squashExpr, true, nil
+}
+
+// getRangeInterval returns the range interval in the range vector node
+// Returns 0 if no range interval is found
+// Example: expression `count_over_time({app="foo"}[10m])` returns 10m
+func getRangeInterval(node parser.Node) time.Duration {
+	switch n := node.(type) {
+	case *parser.AggregateExpr:
+		return getRangeInterval(n.Expr)
+	case *parser.Call:
+		argRangeInterval := time.Duration(0)
+		// Iterate over Call's arguments until a MatrixSelector is found
+		for _, arg := range n.Args {
+			if argRangeInterval = getRangeInterval(arg); argRangeInterval != 0 {
+				break
+			}
+		}
+		return argRangeInterval
+	case *parser.MatrixSelector:
+		return n.Range
+	default:
+		return 0
+	}
+}
+
+// getOffset returns the offset interval in the range vector node
+// Returns 0 if no offset interval is found
+// Example: expression `count_over_time({app="foo"}[10m]) offset 1m` returns 1m
+func getOffset(node parser.Node) time.Duration {
+	switch n := node.(type) {
+	case *parser.AggregateExpr:
+		return getOffset(n.Expr)
+	case *parser.Call:
+		argRangeInterval := time.Duration(0)
+		// Iterate over Call's arguments until a MatrixSelector is found
```

Comment:
Is this comment correct? We actually iterate over Call's argument until we find a `VectorSelector` no, or generally speaking until we find an offset.

Reply:
Yes, correct. Bad copy pasting from the `getRangeInterval` method 😅